### ISSUE TITLE
Processor and OS support AVX512, but GCC doesn't

### ISSUE THREAD

**Issue by thezboe** [jgaeddert/liquid-dsp#331]
It appears that the configure script checks for AVX512 support. In my case, the OS and processor support, but I have an ancient version of GCC that doesn't yet have the AVX512 options. The configure script correctly determines this: 
```
checking whether AVX512-F is supported by the processor... yes
checking whether AVX512-F is supported by the processor and OS... yes
checking whether C compiler accepts -mavx512f... no
configure: WARNING: Your processor and OS supports AVX512-F instructions but not your compiler, can you try another compiler?
checking whether AVX512-CD is supported by the processor... yes
checking whether AVX512-CD is supported by the processor and OS... yes
checking whether C compiler accepts -mavx512cd... no
configure: WARNING: Your processor and OS supports AVX512-CD instructions but not your compiler, can you try another compiler?
checking whether AVX512-PF is supported by the processor... no
checking whether AVX512-ER is supported by the processor... no
checking whether AVX512-VL is supported by the processor... yes
checking whether AVX512-VL is supported by the processor and OS... yes
checking whether C compiler accepts -mavx512vl... no
configure: WARNING: Your processor and OS supports AVX512-VL instructions but not your compiler, can you try another compiler?
checking whether AVX512-BW is supported by the processor... yes
checking whether AVX512-BW is supported by the processor and OS... yes
checking whether C compiler accepts -mavx512bw... no
configure: WARNING: Your processor and OS supports AVX512-BW instructions but not your compiler, can you try another compiler?
checking whether AVX512-DQ is supported by the processor... yes
checking whether AVX512-DQ is supported by the processor and OS... yes
checking whether C compiler accepts -mavx512dq... no
configure: WARNING: Your processor and OS supports AVX512-DQ instructions but not your compiler, can you try another compiler?
checking whether AVX512-IFMA is supported by the processor... no
checking whether AVX512-VBMI is supported by the processor... no
```

But, the resulting makefile still tries to use the flags, resulting in error.
```
gcc -g -O2  -mavx512f  -Wall -fPIC -Wno-deprecated -Wno-deprecated-declarations -I. -Iinclude  -c -o src/libliquid.o src/libliquid.c
gcc: error: unrecognized command line option \u2018-mavx512f\u2019
make: *** [src/libliquid.o] Error 1
```

**Comment by thezboe** [comment]
It appears that configure.ac should be checking `ax_cv_support_avx512f_ext` instead of `ax_cv_have_avx512f_ext`?

**Comment by jgaeddert** [comment]
Good catch. As much as I hate to admit it, this is where CMake is useful...

**Comment by pfeatherstone** [comment]
Have you looked at zigbuild ? Or zig for that matter. It has a nice embedded build system and can cross compile (C, C++ and Zig) to tons of targets using the same binary.

**Comment by JayKickliter** [comment]
> Have you looked at zigbuild ? Or zig for that matter. It has a nice embedded build system and can cross compile (C, C++ and Zig) to tons of targets using the same binary.

Zig is pretty awesome for cross-compiling C, and as much as I hate CMake/autotools, using anything else is huge pain for downstream users.

**Comment by jgaeddert** [comment]
I *have* looked into zig! Supports built-in testing as well!

**Comment by lukeprince20** [comment]
> Good catch. As much as I hate to admit it, this is where CMake is useful...

CMake is great (IMO)! But checking for microarchitectures is still pretty niche. Earlier this year I added https://github.com/jgaeddert/liquid-dsp/pull/314 which added CMake build support for the previous release version. In it I used Google's awesome [cpu_features](https://github.com/google/cpu_features) library for detection of supported CPU instruction sets. You can see how I hooked it into the build configuration [here](https://github.com/lukeprince20/liquid-dsp/blob/cmake_build/CMakeLists.txt#L252-L329). If there's interest I can resurrect this PR for the current release version 1.6.

**Comment by AHSauge** [comment]
Maybe I should create a separate issue for this, but my main problem here is that this is checked at compile time. I can understand that it's convenient, but it makes packaging and distribution complicated. What is packaged in Debian won't run on my build HW for example, because the .deb is built with AVX2. It would be much better to do this at run-time, and pick the best option for the HW. cpu_features mentioned above seems like a good option to incorporate that.

For now I'm just working around this, but if there's sufficient interest I might look into implementing something for this.

**Comment by jgaeddert** [comment]
This is a strong point. Would you suggest having pre-compiled functions for e.g. dot product using various SIMD method (SSE, MMX, AVX, AVX512 etc.) and then using a switch statement to run the one supported by the architecture on the host system? That's probably not a lot of overhead and wouldn't make the library _that_ much larger.

I suppose the build would check all possible versions that are available and compile them all. If someone wanted to avoid run-time overhead, they could possibly narrow what's built using cmake options.

**Comment by AHSauge** [comment]
There are a couple of ways to solve that. I saw FMV mentioned elsewhere, and that might be the easiest way to solve this for Linux. Generally I believe this is solved by using function pointers, so at start-up run some CPUID (or similar) checks to find the most optimal implementation that can be run, and set functions pointer that can then be called afterwards. 

I suspect this is mostly an x86 issue, so I would indeed suggest to make it optional somehow. A pitfall here is that compiling sub-sections with AVX and the library as a whole without it can lead to deteriorated performance, so there's a genuine use case to disable this or compile with a higher minimum than SSE2 (which is implied by x86_64).

**Comment by jgaeddert** [comment]
These changes should properly detect which SIMD instructions are available: aca5ba21a2e3ce95cfc2b2ef24bc5eb713d049f1...ab4c83fdcad0069e18ccc915e808ee1d816ad1b0

I have tested on a few systems (and PPC through emulation) and it seems to work. Next I would need to build _all_ variants within the library and then select the right one at runtime. The only issue with this is that the structured dot product can have unique members depending upon the version used (e.g. altivec has four copies to ensure alignment).
